2023-risks/m10-insufficient-binary-protection.md

```diff
@@ -0,0 +1,102 @@
+
+For a quick check, developers could inspect their own app binaries using similar tools as attackers would use. There are many free or affordable tools, like IDA Pro, Hopper, otool, apktool and Ghidra that are also quite easy to use and well documented.
+
+
+# How Do I Prevent 'Insufficient Binary Protection'?
+
+For each app, it should be assessed whether any critical content is contained in the binary or whether its popularity mandates binary protection. If yes, a threat modeling analysis helps to identify the highest risks and their expected financial impact in case that they occur. For the most relevant risks, countermeasures should be taken. 
+
+As always, apps run in untrusted execution environments and should only get the least neccesary information they need to work, as it is always at risk of being leaked or manipulated. But assuming that certain secrets, algorithms, security checks and similar must be within the app's binary, different attacks can be fent off by different means:
+
+**Reverse engineering:** To prevent reverse engineering, the app binary should be made incomprehensible. This is supported by many free and commercial obfuscation tools. Compiling part of apps natively (iOS and Android) or using interpreters or nested virtual machines makes reverse engineering even harder, as many decompiling tools only support one language and binary format. This kind of obfuscation is a tradeoff between complexity of the code and robustness against reverse engineering, as many libraries that rely on certain strings or symbols in the code wont work with full obfuscation. Developers could check the quality of their obfuscation by using the tools from the previous section.
+
+**Breaking security mechanisms:** Obfuscation also helps against manipulation, as an attacker has to understand the control flow in order to skip security checks and alike. In addition, local security checks should be enforced also by the backend. For example, required resources for a protected feature should only be downloaded if a check succeeds locally *and* in the backend. Finally, integrity checks could detect code tampering and render the app installation unusable, e.g., by deleting some ressources. However, such an integrity check could also be found and deactivated as any other local security check.
+
+**Redistribution** (with malicious code): Integrity checks, e.g., on startup, could also detect redistribution and modification of app binaries. These violations could automatically be reported to find and remove the unauthorized copies of the app from the app stores before they become widespread. There are also specialized companies that support this use case.
```

Comment:
```suggestion
For each app, it should be assessed whether any critical content is contained in the binary or whether its popularity mandates binary protection. If yes, a threat modeling analysis helps to identify the highest risks and their expected financial impact in case they occur. For the most relevant risks, countermeasures should be taken. 

As always, apps run in untrusted execution environments and should only get the least necessary information they need to work, as it is always at risk of being leaked or manipulated. But assuming that certain secrets, algorithms, security checks, and similar must be within the app's binary, different attacks can be fent off by different means:

**Reverse engineering:** To prevent reverse engineering, the app binary should be made incomprehensible. This is supported by many free and commercial obfuscation tools. Compiling part of apps natively (iOS and Android) or using interpreters or nested virtual machines makes reverse engineering even harder, as many decompiling tools only support one language and binary format. This kind of obfuscation is a tradeoff between the complexity of the code and robustness against reverse engineering, as many libraries that rely on certain strings or symbols in the code won't work with full obfuscation. Developers could check the quality of their obfuscation by using the tools from the previous section.

**Breaking security mechanisms:** Obfuscation also helps against manipulation, as an attacker has to understand the control flow in order to skip security checks and like. In addition, local security checks should be enforced also by the backend. For example, required resources for a protected feature should only be downloaded if a check succeeds locally *and* in the backend. Finally, integrity checks could detect code tampering and render the app installation unusable, e.g., by deleting some resources. However, such an integrity check could also be found and deactivated as any other local security check.

**Redistribution** (with malicious code): Integrity checks, e.g., on startup, could also detect redistribution and modification of app binaries. These violations could automatically be reported to find and remove the unauthorized copies of the app from the app stores before they become widespread. There are also specialized companies that support this use case.
```